File: Source/DotNET/Backend/GraphQL/SchemaRoute.cs
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Collections.Generic;
using HotChocolate.Types;
using Microsoft.AspNetCore.Authorization;

namespace Dolittle.Vanir.Backend.GraphQL
{
    public class SchemaRoute : ObjectType
    {
        readonly List<SchemaRoute> _children = new();
        readonly List<SchemaRouteItem> _items = new();

        public SchemaRoute(string path, string localName, string typeName)
        {
            Path = path;
            LocalName = localName;
            TypeName = typeName;
        }

        public string Path { get; }
        public string LocalName { get; }
        public string TypeName { get; }

        public void AddChild(SchemaRoute child)
        {
            _children.Add(child);
        }

        public void AddItem(SchemaRouteItem item)
        {
            _items.Add(item);
        }

        protected override void Configure(IObjectTypeDescriptor descriptor)
        {
            descriptor.Name(TypeName);

            foreach (var item in _items)
            {
                var fieldDescriptor = descriptor.Field(item.Method).Name(item.Name);

                AddAdornedAuthorization(item, fieldDescriptor);
            }

            foreach (var child in _children)
            {
                descriptor.Field(child.LocalName).Type(child).Resolver(_ => new object());
            }

            if (_items.Count == 0)
            {
                descriptor.Field("Default").Resolve(() => "Configure your first item");
            }
        }

        void AddAdornedAuthorization(SchemaRouteItem item, IObjectFieldDescriptor fieldDescriptor)
        {
            var authorizeAttributes = new List<AuthorizeAttribute>();
            authorizeAttributes.AddRange(item.Method.GetCustomAttributes(typeof(AuthorizeAttribute), true) as AuthorizeAttribute[]);
            authorizeAttributes.AddRange(item.Method.DeclaringType.GetCustomAttributes(typeof(AuthorizeAttribute), true) as AuthorizeAttribute[]);

            foreach (var authorizeAttribute in authorizeAttributes)
            {
                if (string.IsNullOrEmpty(authorizeAttribute.Policy))
                {
                    fieldDescriptor.Authorize();
                }
                else
                {
                    fieldDescriptor.Authorize(authorizeAttribute.Policy);
                }
            }
        }
    }
}

File: Source/typescript/backend/graphql/SchemaRoute.ts
// Copyright (c) Dolittle. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import { GraphQLField, GraphQLFieldConfig, GraphQLFieldConfigMap, argsToArgsConfig } from 'graphql/type/definition';
import { GraphQLObjectType } from 'graphql';
import { PubSub } from 'graphql-subscriptions';

const pubSub = new PubSub();

export class SchemaRoute {
    readonly children: SchemaRoute[] = [];
    readonly items: GraphQLField<any, any>[] = [];


    constructor(readonly path: string, readonly localName: string, readonly typeName: string) { }

    addChild(child: SchemaRoute) {
        this.children.push(child);
    }

    addItem(item: GraphQLField<any, any>) {
        this.items.push(item);
    }

    get isRoot() {
        return this.path === '';
    }

    toGraphQLObjectType(): GraphQLObjectType {
        const fields: GraphQLFieldConfigMap<any, any> = {};
        for (const route of this.children) {
            const type = route.toGraphQLObjectType();
            fields[route.localName] = {
                type,
                resolve: () => []
            };
        }

        for (const item of this.items) {
            fields[item.name] = {
                description: item.description,
                type: item.type,
                args: argsToArgsConfig(item.args),
                resolve: item.resolve,
                subscribe: item.subscribe,
                deprecationReason: item.deprecationReason,
                extensions: item.extensions
            };
        }

        const type = new GraphQLObjectType({
            name: this.typeName,
            fields
        });

        return type;
    }
}
